src/array.vala

```diff
@@ -464,4 +396,175 @@ public class Vast.Array : Object
 
         return val;
     }
+
+    public class Builder
+    {
+        /* The Builder gives a syntax to create a new Array viewing
+         * the current array. 
+         * 
+         * for each dimension we can use
+         *
+         *  this.slice(axis, step, from, to)
+         *
+         * it shall be idential to numpy's slice(from, to, step) indexing syntax.
+         * THRU corresponds to omitted values in Python.
+         * if there is a difference, consider it a bug.
+         * (easier to test once with have GIR working)
+         *
+         * we can also reroute the axis's shape and strides to any axis in the original
+         * array, or set it to a new axis (NEW_AXIS). (implement transpose, e.g.)
+         *
+         *  this.axis(axis, originalaxis)
+         *
+         * if an axis has shape[axis] == 1 or strides[axis] == 0,
+         * then we can change the shape (broadcastable)
+         *
+         *  this.broadcast(axis, newshape)
+         *
+         * the final build array is built after this.end() is called.
+         *
+         * methods can be chained or called in a loop.
+         *
+         */
+        private size_t shape[32];
+        private ssize_t strides[32];
+        private size_t origin;
+        private Array array;
+        private size_t dimension;
+
+        internal Builder(Array array, size_t dimension)
+        {
+            assert (dimension >= array._dimension);
+            for (var i = 0; i < array._dimension; i ++) {
+                shape[i] = array._shape[i];
+                strides[i] = array._strides[i];
+            }
+            for (var i = array._dimension; i < dimension; i ++) {
+                shape[i] = 1;
+                strides[i] = 0;
+            }
+            origin = array.origin;
+            this.array = array;
+            this.dimension = dimension;
+        }
+
+        /* from:to:step */
+        public Builder
+        slice(ssize_t axis, ssize_t from, ssize_t to, ssize_t step=1)
+        {
+            return qslice(axis, (ssize_t[]) &from, (ssize_t[]) &to, step);
+        }
+
+        /* :to:step */
+        public Builder
+        head(ssize_t axis, ssize_t to, ssize_t step=1)
+        {
+            return qslice(axis, null, (ssize_t[]) &to, step);
+        }
+
+        /* from::step */
+        public Builder
+        tail(ssize_t axis, ssize_t from, ssize_t step=1)
+        {
+            return qslice(axis, (ssize_t[]) &from, null, step);
+        }
+
+        /* ::step */
+        public Builder
+        step(ssize_t axis, ssize_t step=1)
+        {
+            return qslice(axis, null, null, step);
+        }
+
+        public Builder
+        qslice(ssize_t axis,
+            [CCode (array_length = false)]
+            ssize_t []? qfrom = null,
+            [CCode (array_length = false)]
+            ssize_t []? qto = null,
+            ssize_t step = 1)
```

Comment:
There would be a lot of duplicated logic if we do it that way and do it right. We will need a internal method that does unprotected slicing -- because slice would wrap the dimension and mess it up (when `to` is -1).

In the old version ( 457a4e4 ) I had a `_slice` method that does unprotected slicing and everyone preprocess the defaults and calls `_slice`. You can see the code is a lot fatter than this version.